plugins/uefi-dbx: snapd integration for UEFI DBX updates (v2)

bboozzoo opened the pull request:
Type of pull request:

- [ ] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [ ] Code fix
- [x] Feature
- [ ] Documentation

This a v2 draft PR (see https://github.com/fwupd/fwupd/pull/7861 for previous proposal) which introduces integration of fwupd with snapd in the context of UEFI DBX updates. The main use case is Ubuntu Core systems, especially those with full-disk encryption (FDE).

In detail, in Ubuntu Core system and hybrid Ubuntu installations using FDE, snapd plays a role in resealing the encryption keys to the TPM. The constructed boot policy includes contents of DBX. In order to have the encryption key unsealed and to keep the system bootable, the DBX update payload must presented to snapd, such that a boot policy will account for the updated content.

The high level idea is for snapd to be notified of a DBX update in progress just before the new content is written to efivars. This allows for the new boot policy to be constructed, with a branch for both the old and new DBX content. This step is referred to as prepare. Subsequent call would happen right after the update completes, enabling snapd to introspect the actual content of DBX and reseal the keys once again taking only the current content into account. This is known as a cleanup step. An additional startup step would be triggered whenever fwupd starts up, such that an update which may have previously been interrupted by a reboot or a crash (whether planned or not) could be accounted for.

Snapd exposes a single integration endpoint via its UNIX socket API at /v2/system-secureboot which accepts JSON requests (as implemented in https://github.com/canonical/snapd/pull/14541).

Initially, we are focusing on a simpler use case of fwupd running within a snap, which addresses the immediate use case of Ubuntu Core systems, but in future, this will be extended to cover the classic distro integration in Ubuntu desktop.

In essence the branch is a v2 attempt to explore means of integration within the structure of fwupd. The previous proposal included a dedicated snapd plugin. However it did introduce significant complexity related to matching of the right devices and hooking into the firmware write path. Hence, this branch explores a simpler approach, by integrating directly within `uefi-dbx` plugin.

Please note that there are some TODOs left and quirks in the code, such as verbose curl and similar debugging aids. I will address all of this before marking the PR as non draft once a consensus as to its shape is reached.

superm1 commented on contrib/snap/snapcraft.yaml:
```diff
@@ -91,6 +91,7 @@ parts:
                        -Dpython=/usr/bin/python3,
                        -Dplugin_flashrom=disabled,
                        -Dplugin_powerd=disabled,
+                       -Dplugin_uefi_dbx_snapd=enabled,
```
Since it's a meson tri state feature with no tests to automatically disable from host or missing deps does this need to be hard enabled?

bboozzoo replied:
Good point, let me look into it.

superm1 commented on libfwupdplugin/fu-snap.c:
```diff
@@ -0,0 +1,27 @@
+gboolean
+fu_snap_is_in_snap(void)
+{
+	if (getenv("SNAP") != NULL) {
```
Maybe just

`return getenv("SNAP") != NULL;`

superm1 commented on plugins/uefi-dbx/meson.build:
```diff
@@ -1,17 +1,32 @@
 cargs = ['-DG_LOG_DOMAIN="FuPluginUefiDbx"']
 plugins += {meson.current_source_dir().split('/')[-1]: true}
 
+if get_option('plugin_uefi_dbx_snapd').require(libcurl.found(),
```
I think you need to cache this in a variable for later use

superm1 commented on plugins/uefi-dbx/meson.build:
```diff
@@ -1,17 +1,32 @@
+]
+uefi_dbx_dependencies = [ plugin_deps ]
+
+if get_option('plugin_uefi_dbx_snapd').allowed()
```
See comment about cached variable.

superm1 commented on plugins/uefi-dbx/fu-uefi-dbx-snapd-notifier.c:
```diff
@@ -0,0 +1,256 @@
+
+	/* TODO support system wide snapd socket for outside of snap scenarios */
+	if (!fu_snap_is_in_snap()) {
+		g_warning("attempted use of snapd notifier outside of snap");
```
This will happen if compiled with snapd support which I think will be relatively common since it's a meson tri state feature gated upon curl and Linux only.

So I think you should downgrade this to "info" level (it won't print then) and return here.

hughsie commented on libfwupdplugin/fu-snap.h:
```diff
@@ -0,0 +1,12 @@
+#include <glib.h>
+
+gboolean
+fu_snap_is_in_snap(void);
```
I don't think we need a whole new file for this; just use `fu-common.c` please.

hughsie commented on plugins/uefi-dbx/fu-uefi-dbx-device.c:
```diff
@@ -227,15 +262,33 @@ fu_uefi_dbx_device_init(FuUefiDbxDevice *self)
+
+	if (self->snapd_notifier != NULL) {
+		g_object_unref(self->snapd_notifier);
+		self->snapd_notifier = NULL;
```
you don't need to do the = NULL thing here, that's only required in `dispose`.

hughsie commented on plugins/uefi-dbx/fu-uefi-dbx-device.c:
```diff
@@ -245,3 +298,33 @@ fu_uefi_dbx_device_new(FuContext *ctx)
+void
+fu_uefi_dbx_device_set_snapd_notifier(FuUefiDbxDevice *self, FuUefiDbxSnapdNotifier *obs)
+{
+	self->snapd_notifier = g_object_ref(obs);
```
g_set_object() here pls.

hughsie commented on plugins/uefi-dbx/fu-uefi-dbx-plugin.c:
```diff
@@ -38,7 +53,30 @@ fu_uefi_dbx_plugin_coldplug(FuPlugin *plugin, FuProgress *progress, GError **err
+
+#ifdef WITH_UEFI_DBX_SNAPD_NOTIFIER
+	/* TODO progress? */
+	if (FU_UEFI_DBX_PLUGIN(plugin)->snapd_notifier != NULL) {
```
nah, just do a `self = FU_UEFI_DBX_PLUGIN(plugin)` at the top of the function pls.

hughsie commented on plugins/uefi-dbx/fu-uefi-dbx-plugin.c:
```diff
@@ -48,19 +86,74 @@ fu_uefi_dbx_plugin_init(FuUefiDbxPlugin *self)
+
+#ifdef WITH_UEFI_DBX_SNAPD_NOTIFIER
+static gboolean
+fu_uefi_dbx_plugin_snapd_notify_init(FuUefiDbxPlugin *self)
```
Could we pass in a GError? I think handling the error in the caller rather than locally would feel more natural.

hughsie commented on plugins/uefi-dbx/fu-uefi-dbx-snapd-notifier.c:
```diff
@@ -0,0 +1,256 @@
+	if (status_code != 200) {
+		if (rsp_buf->len > 0) {
+			/* make sure the response is null terminated */
+			g_byte_array_append(rsp_buf, (const guint8 *)"\0", 1);
```
I think you want fu_strsafe() here

superm1 commented on plugins/uefi-dbx/fu-uefi-dbx-device.c:
```diff
@@ -8,13 +8,26 @@
 
+#ifdef WITH_UEFI_DBX_SNAPD_NOTIFIER
+static gboolean
+fu_uefi_dbx_device_maybe_notify_snapd_prepare(FuUefiDbxDevice *self, GBytes *data, GError **error);
+static gboolean
+fu_uefi_dbx_device_maybe_notify_snapd_cleanup(FuUefiDbxDevice *self, GError **error);
```
could you place the functions before their first use in the file and avoid the prototypes?

superm1 commented:
@hughsie what are your thoughts about all the `#ifdef WITH_UEFI_DBX_SNAPD_NOTIFIER` code?  I'm personally leaning we don't need all those guards, it's not super intrusive.  I'm thinking it's better to just compile it in all the time, and have at most one `#ifdef WITH_UEFI_DBX_SNAPD_NOTIFIER` around whether to use it or not (if not; maybe it should just be conditional on environment by checking `fu_snap_is_snap` and/or looking for the socket)

superm1 commented on plugins/uefi-dbx/fu-uefi-dbx-device.c:
```diff
@@ -33,6 +46,14 @@ fu_uefi_dbx_device_write_firmware(FuDevice *device,
+#ifdef WITH_UEFI_DBX_SNAPD_NOTIFIER
+	g_debug("firmware write - notify snapd");
+	if (!fu_uefi_dbx_device_maybe_notify_snapd_prepare(FU_UEFI_DBX_DEVICE(device), fw, error)) {
+		g_warning("cannot notify snapd of pending update");
```
I think you want `g_prefix_error` here instead of `g_warning`

bboozzoo commented:
Is it ok if I amend relevant patches and force push, or do you prefer patches on top of the current branch?

superm1 commented on plugins/uefi-dbx/fu-uefi-dbx-plugin.c:
```diff
@@ -9,17 +9,32 @@
 
+#ifdef WITH_UEFI_DBX_SNAPD_NOTIFIER
+static gboolean
+fu_uefi_dbx_plugin_snapd_notify_init(FuUefiDbxPlugin *self);
```
just move the function here instead of adding a prototype

superm1 reviewed (commented):
I'm not sure how doable it is due to the external stateful dependencies, but this is a lot of new code without coverage.  Can you think about how you could extend the snap unit tests to add a case for this?

One option is to load the plugin in unit tests and simulate the environment with a runner (rather than an end to end simulation with a real snapd daemon).  This is how the redfish plugin gets unit tests accomplished.

superm1 commented:
> Is it ok if I amend relevant patches and force push, or do you prefer patches on top of the current branch?

Until we have external dependencies on other PRs (which is unlikely for this one) I think it's better to just add fixups for now to make review easier.  It's a simple `git rebase -i origin/main` and then squash them all to one once you're ready to flip from draft.

bboozzoo commented:
> I'm not sure how doable it is due to the external stateful dependencies, but this is a lot of new code without coverage. Can you think about how you could extend the snap unit tests to add a case for this?
> 
> One option is to load the plugin in unit tests and simulate the environment with a runner (rather than an end to end simulation with a real snapd daemon). This is how the redfish plugin gets unit tests accomplished.

Real snapd would be too heavy IMO, but the actual API is very simple. I look into it.

superm1 commented on plugins/uefi-dbx/fu-uefi-dbx-device.c:
```diff
@@ -33,6 +69,14 @@ fu_uefi_dbx_device_write_firmware(FuDevice *device,
 	if (fw == NULL)
 		return FALSE;
 
+#ifdef WITH_UEFI_DBX_SNAPD_NOTIFIER
+	g_debug("firmware write - notify snapd");
+	if (!fu_uefi_dbx_device_maybe_notify_snapd_prepare(FU_UEFI_DBX_DEVICE(device), fw, error)) {
+		g_prefix_error(error, "cannot notify snapd of pending update");
+		return FALSE;
+	}
+#endif
```
shouldn't this be in the prepare firmware callback?  Or is write really the correct place?

bboozzoo replied:
That was another option. The upside of `_write()` is that we're really committed to doing the update at this point. I can try to move that to prepare if it makes more sense. AFAICT the stream held by FuFirmware is seekable, so `fu_firmware_get_bytes()` should work then called in both.

bboozzoo replied:
That's done in `device_cleanup()`, which AFAIU is called eventually when write fails.

However, starting the DBX update operation has some additional side effects on the snapd side, for instance snaps which are part of the boot process (or have artifacts which are measured) will be blocked form updating. Thus, the cleanup is what would be executed in the 'happy' path. Unexpected reboots or a crash in fwupd should result in `fu_uefi_dbx_snapd_notifier_dbx_manager_startup` being called at some point, which triggers an error handling path if an update was started and not finished. If all fails, there's incomplete work objects in snapd (aka. changes, dbx update is tracked as one) get aborted after a generous waiting time, which would unblock everything again. Obviously, once `prepare` was invoked, a subsequent `prepare` will result in an error, if cleanup/startup wasn't called in between.

superm1 commented on plugins/uefi-dbx/fu-uefi-dbx-device.c:
```diff
@@ -227,15 +285,32 @@ fu_uefi_dbx_device_init(FuUefiDbxDevice *self)
+#ifdef WITH_UEFI_DBX_SNAPD_NOTIFIER
+	FuUefiDbxDevice *self = FU_UEFI_DBX_DEVICE(object);
+
+	if (self->snapd_notifier != NULL) {
+		g_object_unref(self->snapd_notifier);
+	}
```
nit: you can drop the curly braces for a one line block.

superm1 commented on plugins/uefi-dbx/fu-uefi-dbx-plugin.c:
```diff
@@ -38,7 +49,28 @@ fu_uefi_dbx_plugin_coldplug(FuPlugin *plugin, FuProgress *progress, GError **err
+		g_debug("adding snapd notifier");
+		fu_uefi_dbx_device_set_snapd_notifier(device, self->snapd_notifier);
+	} else if (!inhibited && self->snapd_integration_supported && fu_snap_is_in_snap()) {
+		/* we're running inside a snap, the device is not inhibited and snapd
+		 * supports integration, in which case this is a hard error and we
+		 * should not give an option to dbx */
```
if you're running in the snap even on an old snapd without the integration, do you want to inhibit perhaps?  It sounds like it's not a good scenario for updating DBX in that case too I'd think.

bboozzoo replied:
I think it should only be made a hard-error only if Ubuntu FDE is detected. I left a TODO about it in `fu_common_check_full_disk_encryption()`, which is hopefully the next bit that I may be working on after finishing the dbx work in fwupd and snapd.

superm1 commented on plugins/uefi-dbx/fu-uefi-dbx-plugin.c:
```diff
@@ -48,19 +80,73 @@ fu_uefi_dbx_plugin_init(FuUefiDbxPlugin *self)
+{
+	g_autoptr(FuUefiDbxSnapdNotifier) obs = fu_uefi_dbx_snapd_notifier_new();
+
+	if (fu_uefi_dbx_snapd_notifier_dbx_manager_startup(obs, error)) {
+		g_set_object(&self->snapd_notifier, obs);
+		return TRUE;
+	}
+
+	return FALSE;
```
need to set an error when returning `FALSE`.

superm1 replied:
can you please invert?  We usually do:

```c
if (!foo(bar, error)
        return FALSE;

return TRUE:
```

superm1 commented on plugins/uefi-dbx/fu-uefi-dbx-plugin.c:
```diff
@@ -48,19 +80,73 @@ fu_uefi_dbx_plugin_init(FuUefiDbxPlugin *self)
+	if (fu_snap_is_in_snap()) {
+		g_autoptr(GError) error_local = NULL;
+		/* only enable snapd integration if running inside a snap */
+		if (fu_uefi_dbx_plugin_snapd_notify_init(FU_UEFI_DBX_PLUGIN(obj), &error_local)) {
+			g_info("snapd integration enabled ");
+			self->snapd_integration_supported = TRUE;
+		} else {
+			/* specific error code if relevant APIs are not present and thus
+			 * integration cannot be supported */
+			self->snapd_integration_supported =
+			    !g_error_matches(error_local, FWUPD_ERROR, FWUPD_ERROR_NOT_SUPPORTED);
+
+			g_info("snapd integration non-functional: %s", error_local->message);
```
would you mind inverting like the above comment here too?

```c
if (!foo(bar, error)
        return FALSE;

return TRUE:
```

superm1 commented on plugins/uefi-dbx/meson.build:
```diff
@@ -1,17 +1,33 @@
 cargs = ['-DG_LOG_DOMAIN="FuPluginUefiDbx"']
 plugins += {meson.current_source_dir().split('/')[-1]: true}
 
+snapd_opt = get_option('plugin_uefi_dbx_snapd')
+
+if snapd_opt.require(libcurl.found(),
+  error_message: 'curl is needed for plugin_uefi_dbx_snapd').disable_auto_if(host_machine.system() != 'linux').allowed()
+endif
```
I think you've got the logic wrong here still because the allowed isn't stored to `snapd_opt` when compiling for non-Linux.

I _think_ this is what you're looking for:

```c
snapd_opt =get_option('plugin_uefi_dbx_snapd').disable_auto_if(host_machine.system() != 'linux').allowed()
if snapd_opt.require(libcurl.found(),
  error_message: 'curl is needed for plugin_uefi_dbx_snapd').allowed()
endif
```

bboozzoo replied:
hm `allowed()` returns `bool`, is this what you meant?
```meson
snapd_opt =get_option('plugin_uefi_dbx_snapd').disable_auto_if(host_machine.system() != 'linux')
if snapd_opt.require(libcurl.found(),
  error_message: 'curl is needed for plugin_uefi_dbx_snapd').allowed()
endif
...
if snapd_opt.allowed()
...
```

superm1 replied:
Yeah.

Hmm how about just optimize then like this to only have the one allowed check? 

```
snapd_opt =get_option('plugin_uefi_dbx_snapd').disable_auto_if(host_machine.system() != 'linux')
if snapd_opt.require(libcurl.found(),
  error_message: 'curl is needed for plugin_uefi_dbx_snapd').allowed()
...
endif # snapd_opt.require().allowed()
```

bboozzoo replied:
There's a separate question whether this should be `.allowed()` or to support auto or rather `.enabled()`

superm1 replied:
> There's a separate question whether this should be .allowed() or to support auto or rather .enabled()

Well we could also sidestep this whole thing by dropping the meson option.  I'm of the opinion that we should just detect snapd automatically in the uefi-dbx plugin, no need to make it a compile option.

@hughsie thoughts on this though?

hughsie replied:
I think I'd rather it be simpler -- so maybe we drop the option?

bboozzoo replied:
> I think I'd rather it be simpler -- so maybe we drop the option?

and make curl a hard requirement?

superm1 commented on plugins/uefi-dbx/fu-uefi-dbx-device.c:
```diff
@@ -8,13 +8,43 @@
+static gboolean
+fu_uefi_dbx_device_maybe_notify_snapd_prepare(FuUefiDbxDevice *self, GBytes *data, GError **error)
+{
+	if (self->snapd_notifier == NULL) {
+		/* nothing to do */
+		return TRUE;
+	}
```
just a nit, can drop the curly braces and the comment

superm1 commented on plugins/uefi-dbx/fu-uefi-dbx-device.c:
```diff
@@ -8,13 +8,43 @@
+static gboolean
+fu_uefi_dbx_device_maybe_notify_snapd_cleanup(FuUefiDbxDevice *self, GError **error)
+{
+	if (self->snapd_notifier == NULL) {
+		/* nothing to do */
+		return TRUE;
+	}
```
just a nit, can drop the curly braces and the comment

superm1 commented on plugins/uefi-dbx/meson.build:
```diff
@@ -1,17 +1,23 @@
+  'fu-uefi-dbx-common.c',
+  'fu-uefi-dbx-device.c',
+  'fu-uefi-dbx-snapd-notifier.c',
+]
+
 plugin_quirks += files('uefi-dbx.quirk')
 plugin_builtin_uefi_dbx = static_library('fu_plugin_uefi_dbx',
-  sources: [
-    'fu-uefi-dbx-plugin.c',
-    'fu-uefi-dbx-common.c',
-    'fu-uefi-dbx-device.c',
-  ],
+  sources: uefi_dbx_sources,
```
nit: i'd just change the one line to add the new source file here

superm1 reviewed (commented):
I guess one side effect of not having it configurable is that it that the uefi-dbx plugin has an implicit dependency on libcurl.

I personally think this is fine, but you either need to do:
1. need to make it explicit at the top of `plugins/uefi-dbx/meson.build` so that we can't try to build uefi-dbx with `get_option('curl')` disabled.
2. Use `#ifdef HAVE_LIBCURL` in the code in applicable places so that it doesn't fail to compile.

@hughsie can you please comment on which way you prefer?  
1. Meson option for the integration?
2. uefi-dbx has hard requirement on libcurl?
3. uefi-dbx has `#ifdef HAVE_LIBCURL`  sprinkled in it.

bboozzoo commented:
> > Is it ok if I amend relevant patches and force push, or do you prefer patches on top of the current branch?
> 
> Until we have external dependencies on other PRs (which is unlikely for this one) I think it's better to just add fixups for now to make review easier. It's a simple `git rebase -i origin/main` and then squash them all to one once you're ready to flip from draft.

@superm1 please have a look at the latest set of changes. I have started adding/extending unit tests for uefi-dbx, I'd like to get some feedback before I commit more time to it. There are obviously some quirks when trying to inspect state or mock things. I would appreciate if you have any suggestions there.

bboozzoo commented on plugins/uefi-dbx/fu-self-test.c:
```diff
@@ -6,7 +6,16 @@
 #include "fu-uefi-dbx-common.h"
+#include "fu-uefi-dbx-plugin.h"
+#include "fu-uefi-dbx-self-test.h"
+#include "glib.h"
```
duh, emacs & clangd
```suggestion
```

bboozzoo commented on plugins/uefi-dbx/fu-self-test.c:
```diff
@@ -45,17 +51,326 @@ fu_efi_image_func(void)
+		   fu_uefi_dbx_test_plugin_startup,
+		   fu_self_test_tear_down);
+
+	g_test_add("/uefi-dbx/coldplug/no-dev",
+		   FuTestFixture,
+		   &simple,
+		   fu_self_test_set_up,
+		   fu_uefi_dbx_test_plugin_coldplug_no_device,
+		   fu_self_test_tear_down);
+
+	g_test_add("/uefi-dbx/coldplug/with-dev",
+		   FuTestFixture,
+		   &simple,
+		   fu_self_test_set_up,
+		   fu_uefi_dbx_test_plugin_coldplug_probed_device,
+		   fu_self_test_tear_down);
+
```
Some of the tests here could be proposed in a separate PR as they improve overall test coverage of the plugin. Let me know if that's preferred and I'll split out the changes.

superm1 replied:
I think leaving them together is totally fine, they're great thanks!

superm1 commented on plugins/uefi-dbx/fu-uefi-dbx-plugin.c:
```diff
@@ -8,9 +8,16 @@
 #include "fu-uefi-dbx-device.h"
 #include "fu-uefi-dbx-plugin.h"
+#include "fu-uefi-dbx-snapd-notifier.h"
+#ifdef FU_UEFI_DBX_SELF_TEST
```
I don't see a need for the `#ifdef`

superm1 commented on plugins/uefi-dbx/fu-uefi-dbx-snapd-notifier.c:
```diff
@@ -0,0 +1,272 @@
+
+#ifdef FU_UEFI_DBX_SELF_TEST
+const char *
+fu_uefi_dbx_self_test_snapd_notifier_replace_socket_path(const char *path) /* nocheck:name */
```
rather than this function, how about sending it from an env variable?

bboozzoo replied:
changed the tests to use FWUPD_SNAPD_SNAP_SOCKET instead

superm1 commented:
> > > Is it ok if I amend relevant patches and force push, or do you prefer patches on top of the current branch?
> > 
> > 
> > Until we have external dependencies on other PRs (which is unlikely for this one) I think it's better to just add fixups for now to make review easier. It's a simple `git rebase -i origin/main` and then squash them all to one once you're ready to flip from draft.
> 
> @superm1 please have a look at the latest set of changes. I have started adding/extending unit tests for uefi-dbx, I'd like to get some feedback before I commit more time to it. There are obviously some quirks when trying to inspect state or mock things. I would appreciate if you have any suggestions there.

I think directionally I like where you're going with this.  The two fundamental things I'm not really happy with:
1. Static globals and special functions used by the test harness only.  I think you can avoid these by using environment variables unique to the plugin and test harness.
2. `ifdef` to compile the plugin code differently in the two context.

At this point, I'd say squash everything down.  You're getting close to something mergable.

superm1 commented on plugins/uefi-dbx/fu-uefi-dbx-plugin.c:
```diff
@@ -48,19 +72,67 @@ fu_uefi_dbx_plugin_init(FuUefiDbxPlugin *self)
+			g_info("snapd integration enabled ");
+			self->snapd_integration_supported = TRUE;
+		}
+	} else {
+		/* TODO figure out non-snap scenarios */
+		g_info("snapd integration outside of snap is not supported");
+	}
```
really you're talking about 3 cases AFAICT.
1. fwupd in a snap
2. fwupd outside of a snap on a system without snapd
3. fwupd outside of a snap on a system with snapd.

Shouldn't that message only show up on case 3?  Would it be harmful to send a "ping" type of message with `fu_uefi_dbx_snapd_notifier_simple_req` in this case to an endpoint?  Maybe a get version command?  I'd expect with a short timeout it wouldn't matter for most systems.

bboozzoo replied:
I think this is mostly correct, but the problem is that cases 2 and 3 may be indistinguishable if the snapd socket isn't available at a given instant when a request is made. In theory snapd ships with a systemd socket unit, which should lead to the service activation when accessed. However, there may be scenarios when there is a significant delay before response is received, or the socket is briefly gone (eg. snapd package is being updated, package manager briefly stops the socket unit).

To keep things simple, case 1 is what I'm focusing on right now, since it's relatively simple and we already know as a fact that if fwupd is inside a snap, snapd must be around as well.

Ideally, as I mentioned in the other PR, the main discriminator here would be presence of Ubuntu-like-FDE, which is detected by inspecting volumes and partitions, making it completely independent of snapd socket presence. Much like with bitlocker and the way fwupd shows additional warning that updating DBX may result in needing to use recovery keys. With proper detection in place, the snapd integration gets simplified to 2 cases, with or without Ubuntu-FDE. The goal is to make integration mandatory iff FDE is present.

In essence, I would like for this to not have any impact on folks who aren't part of the use case, be sufficiently robust for folks who are and finally for myself to not propose a PR with 3kloc added diff :smile: which would obviously make reviewing much more time consuming. Hope this makes sense.

superm1 replied:
Yeah that makes sense.  I think what you're doing for now is fine.

bboozzoo commented on plugins/uefi-dbx/fu-uefi-dbx-snapd-notifier.c:
```diff
@@ -0,0 +1,261 @@
+			rsp = fu_strsafe((const char *)rsp_buf->data, rsp_buf->len + 1);
+		}
+
+		/* TODO check whether the response is even printable? */
+		g_message("snapd request failed with status %ld, response: %s",
+			  (glong)status_code,
+			  rsp != NULL ? rsp : "<none>");
```
Actually it may make sense to keep this one as it shows the exact error response and it only appears if the status is really something else than 200 or 404.

bboozzoo replied:
changed to info

bboozzoo commented:
I've pushed a bunch of test updates. The coverage seems pretty ok at this point:
```
plugins/uefi-dbx/fu-self-test.c              327     318    97%   49-51,53-56,405-406
plugins/uefi-dbx/fu-uefi-dbx-common.c         30       0     0%   12,14-18,22,28-30,33-36,40-43,49,53,57,62,64,68-72,78,80
plugins/uefi-dbx/fu-uefi-dbx-device.c        154     123    79%   60,80,104,106,141-142,147-148,152,155,161,181,191,205-206,213,215-221,223,226,228-233
plugins/uefi-dbx/fu-uefi-dbx-device.h          1       1   100%
plugins/uefi-dbx/fu-uefi-dbx-plugin.c         62      59    95%   40,44,47
plugins/uefi-dbx/fu-uefi-dbx-plugin.h          1       1   100%
plugins/uefi-dbx/fu-uefi-dbx-snapd-notifier.c         82      79    96%   39,110,115
plugins/uefi-dbx/fu-uefi-dbx-snapd-notifier.h          1       1   100%
```

I don't think I'll find time to work on this over the break, but I'm happy to push patches when I'm back in January

bboozzoo commented on plugins/uefi-dbx/fu-uefi-dbx-snapd-notifier.c:
```diff
@@ -0,0 +1,261 @@
+						"}",
+						b64data);
+
+	g_debug("snapd notify prepare, with %" G_GSIZE_FORMAT " bytes of data", bufsz);
```
dropped of debug messages

bboozzoo commented on plugins/uefi-dbx/fu-uefi-dbx-snapd-notifier.c:
```diff
@@ -0,0 +1,272 @@
+
+G_DEFINE_TYPE(FuUefiDbxSnapdNotifier, fu_uefi_dbx_snapd_notifier, G_TYPE_OBJECT)
+
+static const char *snapd_snap_socket = "/run/snapd-snap.socket";
```
dropped

hughsie commented on plugins/uefi-dbx/fu-self-test.c:
```diff
@@ -6,7 +6,19 @@
+#include "fu-plugin-private.h"
+#include "fu-uefi-dbx-device.h"
+#include "fu-uefi-dbx-plugin.h"
+#include "glib.h"
```
I don't think we want/need glib.h here

hughsie commented on plugins/uefi-dbx/fu-self-test.c:
```diff
@@ -6,7 +6,19 @@
+#include "fwupd-error.h"
+
+#include "fu-context-private.h"
+#include "fu-device-private.h"
+#include "fu-device.h"
```
if you've got fu-device-private you get fu-device "for free"

hughsie commented on plugins/uefi-dbx/fu-self-test.c:
```diff
@@ -6,7 +6,19 @@
+#include "fu-uefi-dbx-device.h"
+#include "fu-uefi-dbx-plugin.h"
+#include "glib.h"
+#include "gobject/gmarshal.h"
```
do you need this?

hughsie commented on plugins/uefi-dbx/fu-self-test.c:
```diff
@@ -45,17 +57,636 @@ fu_efi_image_func(void)
+fu_self_test_set_up(FuTestFixture *fixture, gconstpointer user_data)
+{
+	FuTestCase *tc = (FuTestCase *)user_data;
+	g_autoptr(GError) error = NULL;
```
the g_autoptr's are normally last in the variable block pls

hughsie commented on plugins/uefi-dbx/fu-self-test.c:
```diff
@@ -45,17 +57,636 @@ fu_efi_image_func(void)
+{
+	FuTestCase *tc = (FuTestCase *)user_data;
+	g_autoptr(GError) error = NULL;
+	gboolean ret = FALSE;
```
doesn't need `= FALSE`;

hughsie commented on plugins/uefi-dbx/fu-self-test.c:
```diff
@@ -45,17 +57,636 @@ fu_efi_image_func(void)
+	g_autoptr(GError) error = NULL;
+	GPtrArray *devs = NULL;
+	FuDevice *dev = NULL;
+	g_autoptr(GInputStream) stream_fw = NULL;
+	g_autoptr(FuProgress) progress = fu_progress_new(G_STRLOC);
+	g_autoptr(FuPlugin) plugin = fu_plugin_new_from_gtype(fu_uefi_dbx_plugin_get_type(), ctx);
```
same, g_autoptrs last and in case order pls.

hughsie commented on plugins/uefi-dbx/fu-uefi-dbx-device.c:
```diff
@@ -33,6 +59,10 @@ fu_uefi_dbx_device_write_firmware(FuDevice *device,
 	if (fw == NULL)
 		return FALSE;
 
+	if (!fu_uefi_dbx_device_maybe_notify_snapd_prepare(FU_UEFI_DBX_DEVICE(device), fw, error)) {
+		return FALSE;
+	}
```
no need for {} here

hughsie commented on plugins/uefi-dbx/fu-uefi-dbx-snapd-notifier.c:
```diff
@@ -0,0 +1,245 @@
+			    FWUPD_ERROR_INTERNAL,
+			    "failed to communicate with snapd: %s",
+			    curl_easy_strerror(res));
+		return FALSE;
```
I think you're leaking `curl` here -- it might be better to use G_DEFINE_AUTOPTR_CLEANUP_FUNC for the `CURL` like we have elsewhere in fwupd.

hughsie commented on plugins/uefi-dbx/fu-uefi-dbx-snapd-notifier.c:
```diff
@@ -0,0 +1,245 @@
+gboolean
+fu_uefi_dbx_snapd_notifier_dbx_manager_startup(FuUefiDbxSnapdNotifier *self, GError **error)
+{
+	const char *startup_msg = "{"
+				  "\"action\":\"efi-secureboot-update-startup\""
+				  "}";
```
I think one line would be fine, no?

hughsie commented on plugins/uefi-dbx/meson.build:
```diff
@@ -7,15 +9,28 @@ plugin_builtin_uefi_dbx = static_library('fu_plugin_uefi_dbx',
+  install_data(['tests/dbx-d719b2cb-3d3a-4596-a3bc-dad00e67656f'],
+    install_dir: join_paths(installed_test_datadir, 'tests'))
+  install_data(['tests/KEK-8be4df61-93ca-11d2-aa0d-00e098032b8c'],
+    install_dir: join_paths(installed_test_datadir, 'tests'))
```
I'm not a huge fan of adding binary tests blobs into the repo since the whole libxz supply chain issue. Can we build these files using a little python program generator instead?

bboozzoo replied:
I'm not sure that's feasible. The content of dbx and KEK were pulled directly from efivars where I previously enrolled locally generated keys. For some reason the parser wasn't happy with pristine esl neither with self signed ones which I generated with efitools. So at best, if I somehow figure out why the files weren't parsed successfully, the tests will require efitools and some scripting to generate keys on the fly.

bboozzoo replied:
opened https://github.com/fwupd/fwupd-test-firmware/pull/3

superm1 commented:
I just triggered CI again for you.  For the arch build issue it should clean up after a rebase to main.  I'd say go ahead and squash/rebase.  I think you're probably ready to drop the draft tag too soon right?

bboozzoo commented:
Thanks for the reviews. I've squashed the fixups, dropped curl verbose, applied some minor tweaks and pushed. There's a separate PR to fwupd-test-firmware, which adds the blobs needed by unit tests.

bboozzoo commented:
From debian logs: https://github.com/fwupd/fwupd/actions/runs/12709704696/job/35431730837?pr=8210#step:5:2196

```
 FuPluginUefiDbx:ERROR:../plugins/uefi-dbx/fu-self-test.c:276:fu_test_mock_efivar_content: assertion failed (error == NULL): Failed to open file ?/github/workspace/build/plugins/uefi-dbx/tests/KEK-8be4df61-93ca-11d2-aa0d-00e098032b8c?: No such file or directory (g-file-error-quark, 4)
```

the CI may need to be restarted to pick up the branch that was merged to fwupd-test-firmware?

superm1 commented:
(will merge it after we can see tests pass with it missing, and then rerun CI after merging)

bboozzoo commented:
> The new test firmware needs to be optional - skip tests that need it if it's missing.

got it, I'll update the branch accordingly

superm1 commented on plugins/uefi-dbx/fu-self-test.c:
```diff
@@ -45,17 +54,658 @@ fu_efi_image_func(void)
+			g_test_skip("test assets unavailable");
+			return;
+		}
+		g_assert_no_error(error);
```
I think you want this assertion outside of the failure case (you want to make sure that no errors were asserted if `fu_test_mock_dbx_efivars` returned `TRUE`.

superm1 commented on plugins/uefi-dbx/fu-self-test.c:
```diff
@@ -45,17 +54,658 @@ fu_efi_image_func(void)
+			g_test_skip("test assets unavailable");
+			return;
+		}
+		g_assert_no_error(error);
```
same comment as other; move assertion outside failure

superm1 reviewed (approved):
I'm happy with the changes now, thanks.

hughsie commented:
I'm a bit confused why it's not being counted by coverage. Do we need to add it to data/tests/fwupd.sh perhaps?

superm1 commented:
> I'm a bit confused why it's not being counted by coverage. Do we need to add it to data/tests/fwupd.sh perhaps?

We don't have coverage in the Arch job, which is the job that we use for launching test environment scripts.  I tried to get coverage working before in arch and failed so decided Debian builds were "good enough" for now.

superm1 commented:
> > I'm a bit confused why it's not being counted by coverage. Do we need to add it to data/tests/fwupd.sh perhaps?
> 
> We don't have coverage in the Arch job, which is the job that we use for launching test environment scripts. I tried to get coverage working before in arch and failed so decided Debian builds were "good enough" for now.

Hmm, that being said it looks like the job _did make a coverage artifact_.  So maybe something on the underlying code got sorted without any of our doing.

https://github.com/fwupd/fwupd/actions/runs/12712705812/artifacts/2413859284

Anyways, hopefully this fixes it.

https://github.com/fwupd/fwupd/pull/8309

superm1 commented:
> > > I'm a bit confused why it's not being counted by coverage. Do we need to add it to data/tests/fwupd.sh perhaps?
> > 
> > 
> > We don't have coverage in the Arch job, which is the job that we use for launching test environment scripts. I tried to get coverage working before in arch and failed so decided Debian builds were "good enough" for now.
> 
> Hmm, that being said it looks like the job _did make a coverage artifact_. So maybe something on the underlying code got sorted without any of our doing.
> 
> https://github.com/fwupd/fwupd/actions/runs/12712705812/artifacts/2413859284
> 
> Anyways, hopefully this fixes it.
> 
> #8309

I rebased the branch on that (now merged) PR. @hughsie can you look again?

bboozzoo commented:
@superm1 thank you for rebasing the branch!

hughsie reviewed (approved):
LGTM, thanks. @superm1 up to you if you want to include this in 2.0.4 -- there's a new dbx dropping literally today so we can test this on a snap system to see if it all works.

bboozzoo commented:
@superm1 @hughsie thank you!
